File: app/src/main/java/knf/kuma/explorer/FragmentChapters.java
package knf.kuma.explorer;

import android.os.Bundle;
import android.os.Handler;
import android.os.Looper;
import android.preference.PreferenceManager;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ProgressBar;

import androidx.annotation.LayoutRes;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.recyclerview.widget.RecyclerView;
import butterknife.BindView;
import butterknife.ButterKnife;
import knf.kuma.R;
import knf.kuma.database.CacheDB;
import knf.kuma.pojos.ExplorerObject;
import xdroid.toaster.Toaster;

public class FragmentChapters extends Fragment {
    public static final String TAG = "Chapters";
    @BindView(R.id.recycler)
    RecyclerView recyclerView;
    @BindView(R.id.progress)
    ProgressBar progressBar;
    ExplorerChapsAdapter adapter;
    private ClearInterface clearInterface;
    private boolean isFirst = true;

    public FragmentChapters() {
    }

    public static FragmentChapters get(ClearInterface clearInterface) {
        FragmentChapters fragmentChapters = new FragmentChapters();
        fragmentChapters.setInterface(clearInterface);
        return fragmentChapters;
    }

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        View view = inflater.inflate(getLayout(), container, false);
        ButterKnife.bind(this, view);
        return view;
    }

    @LayoutRes
    private int getLayout() {
        if (PreferenceManager.getDefaultSharedPreferences(getContext()).getString("lay_type", "0").equals("0")) {
            return R.layout.recycler_explorer_chaps;
        } else {
            return R.layout.recycler_explorer_chaps_grid;
        }
    }

    public void setObject(ExplorerObject object) {
        clear();
        object.getLiveData(getContext()).observe(this, fileDownObjs -> {
            if (fileDownObjs.size() == 0) {
                Toaster.toast("Directorio vacio");
                CacheDB.INSTANCE.explorerDAO().delete(object);
                clearInterface.onClear();
            } else {
                object.chapters = fileDownObjs;
                progressBar.setVisibility(View.GONE);
                adapter = new ExplorerChapsAdapter(FragmentChapters.this, object, clearInterface);
                recyclerView.setAdapter(adapter);
                if (isFirst) {
                    isFirst = false;
                    recyclerView.scheduleLayoutAnimation();
                }
            }
            //object.clearLiveData(this);
        });
    }

    void deleteAll() {
        if (adapter != null)
            adapter.deleteAll();
    }

    private void clear() {
        isFirst = true;
        adapter = null;
        new Handler(Looper.getMainLooper()).post(() -> {
            if (progressBar != null)
                progressBar.setVisibility(View.VISIBLE);
            if (recyclerView != null)
                recyclerView.setAdapter(null);
        });
    }

    public void setInterface(ClearInterface clearInterface) {
        this.clearInterface = clearInterface;
        if (adapter != null)
            adapter.setInterface(clearInterface);
    }

    public interface ClearInterface {
        void onClear();
    }
}

File: app/src/main/java/knf/kuma/explorer/FragmentChapters.kt
package knf.kuma.explorer

import android.annotation.SuppressLint
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.FrameLayout
import android.widget.ProgressBar
import androidx.annotation.LayoutRes
import androidx.fragment.app.Fragment
import androidx.lifecycle.Observer
import androidx.lifecycle.lifecycleScope
import androidx.recyclerview.widget.RecyclerView
import com.google.android.material.floatingactionbutton.FloatingActionButton
import knf.kuma.R
import knf.kuma.ads.AdsType
import knf.kuma.ads.implBanner
import knf.kuma.backup.firestore.syncData
import knf.kuma.commons.*
import knf.kuma.database.CacheDB
import knf.kuma.pojos.ExplorerObject
import knf.kuma.pojos.RecordObject
import knf.kuma.queue.QueueManager
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import org.jetbrains.anko.find
import org.jetbrains.anko.sdk27.coroutines.onClick
import xdroid.toaster.Toaster


class FragmentChapters : Fragment() {
    lateinit var recyclerView: RecyclerView
    lateinit var progressBar: ProgressBar
    lateinit var fab: FloatingActionButton
    internal var adapter: ExplorerChapsAdapter? = null
    private var clearInterface: ClearInterface? = null
    private var isFirst = true

    private val layout: Int
        @LayoutRes
        get() = if (PrefsUtil.layType == "0") {
            R.layout.recycler_explorer_chaps
        } else {
            R.layout.recycler_explorer_chaps_grid
        }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        val view = inflater.inflate(layout, container, false)
        recyclerView = view.find(R.id.recycler)
        recyclerView.verifyManager(170)
        progressBar = view.find(R.id.progress)
        fab = view.find(R.id.fab)
        view.find<FrameLayout>(R.id.adContainer).implBanner(AdsType.EXPLORER_BANNER, true)
        return view
    }

    private fun playAll(list: List<ExplorerObject.FileDownObj>) {
        lifecycleScope.launch(Dispatchers.Main){
            noCrashSuspend {
                withContext(Dispatchers.IO) {
                    CacheDB.INSTANCE.recordsDAO().add(RecordObject.fromDownloaded(list.last()))
                    syncData { history() }
                    QueueManager.startQueueDownloaded(context, list)
                }
                adapter?.apply {
                    explorerObject.fileList.forEach {
                        it.isSeen = true
                    }
                    notifyDataSetChanged()
                }
            }
        }
    }

    @SuppressLint("RestrictedApi")
    fun setObject(explorerObject: ExplorerObject?) {
        noCrash {
            fab.internalSetVisibility(View.INVISIBLE, true)
            fab.hide()
        }
        clear()
        explorerObject?.let {
            it.getLiveData(context)
                    .observe(this@FragmentChapters, Observer { fileDownObjs ->
                        if (fileDownObjs.isEmpty()) {
                            Toaster.toast("Directorio vacio")
                            lifecycleScope.launch(Dispatchers.IO){
                                CacheDB.INSTANCE.explorerDAO().delete(explorerObject)
                                launch(Dispatchers.Main) {
                                    clearInterface?.onClear()
                                }
                            }
                        } else {
                            explorerObject.chapters = fileDownObjs as MutableList<ExplorerObject.FileDownObj>
                            lifecycleScope.launch(Dispatchers.Main){
                                progressBar.visibility = View.GONE
                                adapter = ExplorerChapsAdapter(this@FragmentChapters, recyclerView, withContext(Dispatchers.IO) { ExplorerObjectWrap(explorerObject) }, clearInterface)
                                recyclerView.adapter = adapter
                                if (isFirst) {
                                    isFirst = false
                                    recyclerView.scheduleLayoutAnimation()
                                }
                                if (!CastUtil.get().connected()) {
                                    fab.show()
                                    fab.onClick { playAll(fileDownObjs) }
                                }
                            }
                        }
                    })
        }
    }

    internal fun deleteAll() {
        adapter?.deleteAll()
    }

    private fun clear() {
        isFirst = true
        adapter = null
        doOnUI {
            progressBar.visibility = View.VISIBLE
            recyclerView.adapter = null
        }
    }

    fun setInterface(clearInterface: ClearInterface) {
        this.clearInterface = clearInterface
        adapter?.setInterface(clearInterface)
    }

    interface ClearInterface {
        fun onClear()
    }

    companion object {
        const val TAG = "Chapters"

        operator fun get(clearInterface: ClearInterface): FragmentChapters {
            val fragmentChapters = FragmentChapters()
            fragmentChapters.setInterface(clearInterface)
            return fragmentChapters
        }
    }
}
